rosserial_arduino/src/rosserial_arduino/make_libraries_isolated.py

```diff
@@ -0,0 +1,144 @@
+# POSSIBILITY OF SUCH DAMAGE.
+# 
+# Modified by Bhavya Gupta for using config files
+
+THIS_PACKAGE = "rosserial_arduino"
+
+__usage__ = """
+make_libraries_isolated.py generates the Arduino rosserial library files for a
+set of packages and installs them as per the directory structure provided in 
+a yaml config file.
+
+rosrun rosserial_arduino make_libraries.py <yaml config file>
+"""
+
+import rospkg
+import rosserial_client
+from rosserial_client.make_library import *
+import yaml
+import os
+
+# for copying files
+import shutil
+
+ROS_TO_EMBEDDED_TYPES = {
+    'bool'    :   ('bool',              1, PrimitiveDataType, []),
+    'byte'    :   ('int8_t',            1, PrimitiveDataType, []),
+    'int8'    :   ('int8_t',            1, PrimitiveDataType, []),
+    'char'    :   ('uint8_t',           1, PrimitiveDataType, []),
+    'uint8'   :   ('uint8_t',           1, PrimitiveDataType, []),
+    'int16'   :   ('int16_t',           2, PrimitiveDataType, []),
+    'uint16'  :   ('uint16_t',          2, PrimitiveDataType, []),
+    'int32'   :   ('int32_t',           4, PrimitiveDataType, []),
+    'uint32'  :   ('uint32_t',          4, PrimitiveDataType, []),
+    'int64'   :   ('int64_t',           8, PrimitiveDataType, []),
+    'uint64'  :   ('uint64_t',          4, PrimitiveDataType, []),
+    'float32' :   ('float',             4, PrimitiveDataType, []),
+    'float64' :   ('float',             4, AVR_Float64DataType, []),
+    'time'    :   ('ros::Time',         8, TimeDataType, ['ros/time']),
+    'duration':   ('ros::Duration',     8, TimeDataType, ['ros/duration']),
+    'string'  :   ('char*',             0, StringDataType, []),
+    'Header'  :   ('std_msgs::Header',  0, MessageDataType, ['std_msgs/Header'])
+}
+
+# need correct inputs
+if (len(sys.argv) < 2):
+    print __usage__
+    exit()
+    
+# get config file name
+config_file = sys.argv[1]
+fd_in = open(config_file)
+contents = fd_in.read()
+fd_in.close()
+parsed_contents = yaml.load(contents)
+
+microcontroller_ws = parsed_contents["microcontroller_ws"]
+header_out = os.path.join(microcontroller_ws, parsed_contents["header_output_directory"])
+source_out = os.path.join(microcontroller_ws, parsed_contents["source_output_directory"])
+custom_ws = parsed_contents["catkin_workspaces"]
+packages = parsed_contents["packages"]
+
+print ("\033[40;33mMicrocontroller workspace:\033[0m %s " % microcontroller_ws)
+print ("\033[40;33mHeader output            :\033[0m %s " % header_out)
+print ("\033[40;33mSource output            :\033[0m %s " % source_out)
+print ("\033[40;33mCatkin workspaces        :\033[0m %s " % custom_ws)
+print ("\033[40;33mPackages                 :\033[0m %s " % packages)
+print ("----------------------------------------------------------------------")
+
+# "Overlay" the custom catkin workspaces
+ros_package_path = os.environ["ROS_PACKAGE_PATH"]
+for ws in custom_ws:
+    ros_package_path = ros_package_path + ":" + ws
```

Comment:
Well I have to admit that I forgot that the order matters in overlaying while writing this feature. However, this can still be managed with the current framework by stating that the list is ordered in the yaml file. If anything, this scripts makes it much easier to overlay - than calling `source .../devel/setup.bash` for a bunch of directories.

Also you are right - since our `rosserial` package is in the same catkin ws as the messages - that explains why you and I didin't see a problem at runtime.